File: typescript/game-of-life/game-of-life.ts
/**
 * Implements Conway's Game of Life.
 *
 * This class simulates the evolution of a grid of cells based on a set of rules.
 * Each cell can be either alive or dead. The simulation advances in discrete
 * time steps, or "ticks."
 */
export class GameOfLife {
  /**
   * @var _state The current state of the game grid, represented as a matrix of numbers.
   */
  private _state: number[][];

  /**
   * Creates a new GameOfLife instance with a starting state.
   *
   * @param matrix The initial grid state. Alive cells are represented by non-zero numbers.
   */
  constructor(matrix: number[][]) {
    this._state = matrixCopy(matrix);
  }

  /**
   * Advances the game state by one generation.
   *
   * The new state is calculated based on the rules of Conway's Game of Life
   * applied to the current state.
   */
  public tick() {
    this._state = matrixMap(
      this._state,
      (_, x: number, y: number) => this.willAlive(x, y)
    );
  }

  /**
   * Returns a copy of the current state of the game grid.
   *
   * This method ensures that the internal state of the class cannot be
   * modified directly from outside.
   *
   * @returns A copy of the matrix representing the current state.
   */
  public state(): number[][] {
    return matrixCopy(this._state);
  }

  /**
   * Determines if a cell at a given position will be alive in the next generation.
   *
   * This is based on the number of its alive neighbors and the rules of the game.
   *
   * @param x The x-coordinate of the cell.
   * @param y The y-coordinate of the cell.
   * @returns 1 if the cell will be alive, 0 otherwise.
   */
  private willAlive(x: number, y: number): number {
    const aliveNeighboursCount = this.aliveNeighboursCount(x, y);

    if (aliveNeighboursCount === 3) {
      return 1;
    }
    if (aliveNeighboursCount === 2 && this.isAlive(x, y)) {
      return 1;
    }
    return 0;
  }

  /**
   * Counts the number of alive neighbors for a cell at a given position.
   *
   * @param x The x-coordinate of the cell.
   * @param y The y-coordinate of the cell.
   * @returns The count of alive neighbors.
   */
  private aliveNeighboursCount(x: number, y: number): number {
    const infLimitOfX = Math.max(x - 1, 0);
    const supLimitOfX = Math.min(x + 1, this._state.length - 1);
    const infLimitOfY = Math.max(y - 1, 0);
    const supLimitOfY = Math.min(y + 1, this._state.length - 1);

    let count = 0;

    for (let indexOfY = infLimitOfY; indexOfY <= supLimitOfY; indexOfY++) {
      for (let indexOfX = infLimitOfX; indexOfX <= supLimitOfX; indexOfX++) {
        if (indexOfY === y && indexOfX === x) {
          continue;
        }
        if (this._state[indexOfY][indexOfX] !== 0) {
          count++;
        }
      }
    }

    return count;
  }

  /**
   * Checks if a cell at a given position is alive.
   *
   * An alive cell is represented by a non-zero value.
   *
   * @param x The x-coordinate of the cell.
   * @param y The y-coordinate of the cell.
   * @returns True if the cell is alive, false otherwise.
   */
  private isAlive(x: number, y: number): boolean {
    return this._state[y][x] !== 0;
  }
}

/**
 * Creates a shallow copy of a 2D matrix.
 *
 * @template T The type of elements in the matrix.
 * @param matrix The matrix to copy.
 * @returns A new, independent matrix with the same elements.
 */
function matrixCopy<T>(matrix: T[][]): T[][] {
  return matrix.map(array => [...array]);
}

/**
 * Creates a new matrix by applying a mapper function to each element of an existing matrix.
 *
 * @template T The type of elements in the source matrix.
 * @template U The type of elements in the resulting matrix.
 * @param matrix The source matrix.
 * @param mapper The function to apply to each element. It receives the element, x, and y coordinates.
 * @returns A new matrix with the results of the mapper function.
 */
function matrixMap<T, U>(
  matrix: T[][],
  mapper: (element: T, x: number, y: number) => U
): U[][] {
  const result = newSquareMatrix<U>(matrix.length);

  for (let indexOfY = 0; indexOfY < matrix.length; indexOfY++) {
    for (let indexOfX = 0; indexOfX < matrix.length; indexOfX++) {
      result[indexOfY][indexOfX] = mapper(
        matrix[indexOfY][indexOfX],
        indexOfX, indexOfY
      );
    }
  }

  return result;
}

/**
 * Creates a new empty square matrix of a given length.
 *
 * @template T The type of elements the matrix will hold.
 * @param len The length of the sides of the square matrix.
 * @returns An empty square matrix.
 */
function newSquareMatrix<T>(len: number): T[][] {
  const result = [];

  for (let i = 0; i < len; i++) {
    result.push([]);
  }

  return result;
}

File: lua/game-of-life/game-of-life.lua
--- A module for simulating Conway's Game of Life.
local game_of_life <const> = {}

local next_life_state, alive_neighbours_count, neighbours_of, reject

--- A type alias representing the game board, which is a two-dimensional array of `LifeState` values.
--- @alias GameState LifeState[][]
--- An enumeration defining the two possible states for a cell in the game.
--- @enum LifeState
local LifeState <const> = {
    Dead = 0,
    Alive = 1
}

--- Advances the game state by one generation (or "tick").
--- This function calculates the next state for every cell on the board based on the rules of the Game of Life.
--- @param state GameState The current state of the game board.
--- @return GameState -- A new game state representing the next generation.
function game_of_life.tick(state)
    local result = {}

    for y = 1, #state, 1 do
        result[y] = {}
        for x = 1, #state[1], 1 do
            result[y][x] = next_life_state(state, x, y)
        end
    end

    return result
end

--- Determines the next state of a single cell based on the rules of the Game of Life.
--- A dead cell with exactly 3 alive neighbors becomes alive. An alive cell with 2 or 3 alive neighbors remains alive. All other cells die.
--- @param state GameState The current state of the game board.
--- @param x integer The x-coordinate of the cell.
--- @param y integer The y-coordinate of the cell.
--- @return LifeState -- The new state of the cell.
function next_life_state(state, x, y)
    local alive_neighbours = alive_neighbours_count(state, x, y)

    if alive_neighbours == 3 then
        return LifeState.Alive
    elseif state[y][x] == LifeState.Alive and alive_neighbours == 2 then
        return LifeState.Alive
    else
        return LifeState.Dead
    end
end

--- Counts the number of alive neighbors for a specific cell.
--- @param state GameState The current state of the game board.
--- @param x integer The x-coordinate of the cell.
--- @param y integer The y-coordinate of the cell.
--- @return integer -- The number of alive neighbors.
function alive_neighbours_count(state, x, y)
    local count = 0

    for _, neighbour in ipairs(neighbours_of(#state[1], #state, x, y)) do
        if state[neighbour.y][neighbour.x] == LifeState.Alive then
            count = count + 1
        end
    end

    return count
end

--- Returns a table of valid coordinates for all direct neighbors of a cell.
--- The list excludes the cell itself and any coordinates that are outside the board boundaries.
--- @param width integer The width of the game board.
--- @param height integer The height of the game board.
--- @param x integer The x-coordinate of the cell.
--- @param y integer The y-coordinate of the cell.
--- @return {x: integer, y: integer}[] -- A table of neighbor coordinates.
function neighbours_of(width, height, x, y)
    local neighbours = {}

    for i = y - 1, y + 1, 1 do
        for j = x - 1, x + 1, 1 do
            table.insert(neighbours, { x = j, y = i })
        end
    end

    return reject(neighbours, function(coord)
        return (
            coord.x < 1 or coord.y < 1 or
            coord.x > width or coord.y > height or
            (coord.x == x and coord.y == y)
        )
    end)
end

--- A general-purpose helper function that filters a list based on a rejection function.
--- @param list table The list to filter.
--- @param rejecter function The function that returns `true` for elements to be rejected.
--- @return table -- A new list containing only the elements that were not rejected.
function reject(list, rejecter)
    local result = {}

    for _, elem in ipairs(list) do
        if not rejecter(elem) then
            table.insert(result, elem)
        end
    end

    return result
end

return game_of_life